### DISCUSSION TITLE
Add programmatic Comments support for use with ESLint

### DISCUSSION THREAD

**Discussion by vadimofd** [graphql-java/graphql-java#3250] [Ideas]
We use [graphql-eslint](https://github.com/B2o5T/graphql-eslint) for schema linting and frequently use [ESLint configuration comments](https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1) when [disabling rules](https://the-guild.dev/graphql/eslint/docs/getting-started#disabling-rules).

Our schema is defined programmatically and we generate SDL using [SchemaPrinter](https://github.com/graphql-java/graphql-java/blob/v19.2/src/main/java/graphql/schema/idl/SchemaPrinter.java). However, the comments we add in Java code do not get printed to SDL by `SchemaPrinter`.

So, we ended up modifying `SchemaPrinter` to support our programmatic comments. This seems to work well for this purpose and our code changes seem to be relatively simple.

Would like to suggest this as a feature.

PS: I saw similar request previously discussed in [Consider adding support for GraphQL comments to SDL creation](https://github.com/graphql-java/graphql-java/issues/1848) but the schema linting angle was not part that discussion and makes this need more significant than just documentation need.

**Comment by bbakerman** [comment]
This is a tricky thing because according to the graphql specification

https://spec.graphql.org/June2018/#sec-Comments

> Comments behave like white space and may appear after any token, or before a line terminator, and have no significance to the semantic meaning of a GraphQL Document.

So once a document is parsed into a runtime representation (eg the objects like `GraphQLObjectType`), comments have no meaning.

Types and fields and arguments in graphql instead have a description.

https://spec.graphql.org/June2018/#Description

```
"""
A simple GraphQL schema which is well described.
"""
type Query {
  """
  Translates a string from a given language into a different language.
  """
  translate(
    "The original language that `text` is provided in."
    fromLanguage: Language

    "The translated language to be returned."
    toLanguage: Language

    "The text to be translated."
    text: String
  ): String
}
```

This is captured in a first class way when you create the runtime representation of a type or field or argument

eg `graphql.schema.GraphQLObjectType#getDescription`

But this linting package has decided to use `//` comments and these cant be easily round tripped.

Now it turns out is not so simple.  There is a runtime representation of a type AND there is an AST representation of that type and the 2 are related BUT not the same

`graphql.schema.GraphQLObjectType#getDefinition` returns the (ooptional) AST node that defined the runtime representation say

`graphql.language.ObjectTypeDefinition` can (optionally) capture the AST element and its comments via `graphql.language.AbstractNode#getComments` which is a list of comments associated with the AST element

BUT its tricky because imagine this valid type definition

```
// comment 1
" description of Foo"
type Foo {
   fieldA : Boolean
}

// comment 2
extend type Foo {
  fieldB : Boolean
}
```

This results in 1 runtime object type `Foo` but it came from N AST elements, each with their own comments.  So how do they apply to `Foo` - what order etc.?

Well it turns out that `graphql.schema.idl.SchemaPrinter.Options#isUseAstDefinitions` has a boolean (off by default) to use the AST definitions if they are present.   

However in line with the graphql spec - the `AstPrinter` ignores comments - because

>  Comments behave like white space and may appear after any token, or before a line terminator, and have no significance to the semantic meaning of a GraphQL Document.


However....

We do have an experimental class `graphql.language.PrettyAstPrinter` which does keep comments HOWEVER its not used by the SchemaPrinter so you would need to traverse the schema element tree yourself and call `PrettyAstPrinter` on it yourself.
  

If this feels confusing - it is because we have competing demands here - we have a graphql spec saying comments dont matter, we have some people who dont want them stored for memory reasons (including the AST elements) and then we have some like yourself who DO want to capture comments.

So given these design trade offs we decided in the main

* we dont have a place to capture comments onto the runtime schema objects
* we dont want to add a place to capture comments 
   *  because the spec says they are semantically insignificant
   * people (in the past) have complained it burns memory unnecessarily 

We would be interested in seeing how you changed SchemaPrinter to include comments say.

**Reply by vadimofd** [reply]
Thank you very much for such detailed and informative response @bbakerman!

We haven't looked at the `PrettyAstPrinter` yet as I think it was just recently introduced in v20.x while we are on v19.x for now. Will take a look in the next few weeks!

I'll be away on vacation for a bit so will post our `SchemaPrinter` changes once I am back.

**Reply by vadimofd** [reply]
Apologies for the delay. Please, find our customized schema printer based off version 19.2 here
https://gist.github.com/vadimofd/ae208b7f886ac538e72c8c2fe5365ab7

File comments mention names of the few customized methods.

**Reply by bbakerman** [reply]
Ahh interesting code - basically you grab the comments on GraphqlSchema objects from the possibly attached AST definitions

This will work because

```
 public static class Options {
        private final boolean useCommentsAsDescription;
        private final boolean captureAstDefinitions; // is true by default
        private final boolean useAppliedDirectivesOnly;
```

Nice

**Reply by vadimofd** [reply]
Correct. Seems to do the job without significant side effects, I think. What do you think about adding this feature?

**Reply by bbakerman** [reply]
I think this would be a great feature

We would need another option in the schema printer options like you have and I think it would be off by default say

but then the code that I saw would work well I think

Are you proposing to make a PR here?

**Reply by vadimofd** [reply]
Option with off default makes sense.

Yes, I would propose a PR. Not sure of the end-to-end process for this project but I can take a stab at it if you could point me in the right direction.

**Reply by bbakerman** [reply]
>  Not sure of the end-to-end process for this project but I can take a stab at it if you could point me in the right direction.

Simply fork the graphql-java into your own copy and then make the code changes.

You can then raise a PR against graphql-java from that fork

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork


The code you linked is pretty much ready to go and needs to be retrofitted back into SchemaPrinter

**Comment by vadimofd** [comment]
WIP PR #3287
